File: Sumin/MaskInfo_java/app/src/main/java/org/techtown/maskinfo_java/MainViewModel.java
package org.techtown.maskinfo_java;

import android.util.Log;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import org.techtown.maskinfo_java.model.Store;
import org.techtown.maskinfo_java.model.StoreInfo;
import org.techtown.maskinfo_java.repository.MaskService;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.moshi.MoshiConverterFactory;

public class MainViewModel extends ViewModel {
    private static final String TAG = MainViewModel.class.getSimpleName();
    public MutableLiveData<List<Store>> itemLiveData = new MutableLiveData<>();

    //통신
    private Retrofit retrofit = new Retrofit.Builder()
            .baseUrl(MaskService.BASE_URL)
            .addConverterFactory(MoshiConverterFactory.create())
            .build();

    private MaskService service = retrofit.create(MaskService.class);

    private Call<StoreInfo> storeInfoCall = service.fetchStoreInfo();

    public MainViewModel(){
        fetchStoreInfo();
    }

    public void fetchStoreInfo(){
        storeInfoCall.clone().enqueue(new Callback<StoreInfo>() {
            @Override
            public void onResponse(Call<StoreInfo> call, Response<StoreInfo> response) {
                Log.e(TAG, "onResponse: refresh");
                List<Store> items = response.body().getStores()
                        .stream()
                        .filter(item -> item.getRemainStat() != null)
                        .collect(Collectors.toList());

                itemLiveData.postValue(items);
            }

            @Override
            public void onFailure(Call<StoreInfo> call, Throwable t) {
                Log.e(TAG, "onFailure: ", t);
                itemLiveData.postValue(Collections.emptyList());
            }
        });
    }
}

File: mask_kotlin/app/src/main/java/org/techtown/mask_kotlin/MainViewModel.kt
package org.techtown.mask_kotlin

import android.location.Location
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.launch
import org.techtown.mask_kotlin.model.Store
import org.techtown.mask_kotlin.repository.MaskService
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory

class MainViewModel : ViewModel() {
    val itemLiveData = MutableLiveData<List<Store>>()
    val loadingLiveData = MutableLiveData<Boolean>()
    var location: Location? = null

    private val service: MaskService

    init {
        val retrofit = Retrofit.Builder()
            .baseUrl(MaskService.BASE_URL)
            .addConverterFactory(MoshiConverterFactory.create())
            .build();

        service = retrofit.create(MaskService::class.java)

        fetchStoreInfo()
    }

    fun fetchStoreInfo(){
        // 로딩 시작
        loadingLiveData.setValue(true);

        viewModelScope.launch {
            val storeInfo = service.fetchStoreInfo(37.54479207942774, 127.05543016589836)
            itemLiveData.value = storeInfo.stores

            // 로딩 끝
            loadingLiveData.value = false;
        }

    }
}